### DISCUSSION TITLE
Using Quartz with multiple databases / multiple schedulers

### DISCUSSION THREAD

**Discussion by janHildebrandt98** [quartznet/quartznet#1486] [Q&A] [answered]
Hello everyone!

we are working on a multi-tenancy system and our architecture involves having one database per tenant. 
There is some logic that we need to execute regularly for **all** tenants and that's where Quartz comes in.

So we want to start one job per tenant and then perform the logic. But then there are 2 issues:
1. The connection string for the persistent store is configured only once for the scheduler. But we would need the Quartz state to be saved to / retrieved from each of the tenant databases individually. Is this possible? How do we set this up?
2. Each job needs an Entity Framework DB context set up with the respective connection string to access the correct database. I guess this could be solved using Dependency Injection scopes?

Thanks in advance for your answers & your time.

**Comment by lahma** [comment, marked as answer]
> 1. The connection string for the persistent store is configured only once for the scheduler. But we would need the Quartz state to be saved to / retrieved from each of the tenant databases individually. Is this possible? How do we set this up?

If you have Quartz tables for each tenant separately, you need to configure as many scheduler instances as there are tenants, scheduler factory configuration takes connection string for persistent store and it's bound per scheduler.

Alternative approach is to have separate database for Quartz state and then configure trigger for each tenant and supply that trigger with relevant information, like job data map: key -> Tenant" , value -> "Customer X". You could also provide the connection string, but might be wiser to only supply tenant id and resolve the correct connection string at runtime.

> 2. Each job needs an Entity Framework DB context set up with the respective connection string to access the correct database. I guess this could be solved using Dependency Injection scopes?

This can be a bit tricky as you probably need to build your own job factory that resolves different DbContext based on target (or job map data like I described above). See [MicrosoftDependencyInjectionJobFactory](https://github.com/quartznet/quartznet/blob/main/src/Quartz.Extensions.DependencyInjection/Simpl/MicrosoftDependencyInjectionJobFactory.cs), it might suffice to override `ConfigureScope`. `TriggerFiredBundle` will give you access to job and trigger data.

**Reply by janHildebrandt98** [reply]
Thanks for you reply! 
We also had the idea of creating a separate databse to host the Quartz tables. Might make things easier.
To be honest, i did not know that one could configure multiple scheduler instances. I thought this would override previous configurations, hence my question.

**Reply by lahma** [reply]
I would recommend separate database with single scheduler if that's possible, like you've anticipated it would make things easier. Just distinguish tenant specific things with job/trigger group name for example.

**Reply by janHildebrandt98** [reply]
...We decided to implement the multiple schedulers approach. I added another answer below, describing what we did.

**Comment by janHildebrandt98** [comment]
We decided to try the one scheduler per tenant approach anyway and got it to work after a lot of thought and debugging.
Sadly, I cannot share the code here due to company restrictions but I will try my best to explain what we did in order to get this to work in case anyone needs it in the future.

As already explained, we basically have a list of tenants and want to start one scheduler to handle the jobs for each tenant and store its state in a tenant specific DB. To achieve this we did the following:

1. Create a singleton scheduler repository class that holds a list of all the schedulers and has an API to start/stop/add/remove them
2. To build a scheduler instance, we use the `SchedulerBuilder`, [example ](https://www.quartz-scheduler.net/documentation/quartz-3.x/quick-start.html#configuration)
3. When creating the instance, you can pass in a `NameValueCollection` which contains configuration for that scheduler and this is where it gets interesting:
For the config keys listed [here](https://www.quartz-scheduler.net/documentation/quartz-3.x/quick-start.html#configuration) you need to replace the name-part with some unique tenant identifier. Quartz creates a static map of connections in the background (where it can store the scheduler/job state) and uses the name-part as key.
So you have to create a connection for each tenant to store its state by setting 
`quartz.dataSource.{tenantId}.connectionString` to the respective connection string and assign this connection to your scheduler by setting `quartz.jobStore.dataSource` to the value of `{tenantId}`.
When you configured everything to your liking and added the jobs, the scheduler is set up.
4. Create an implementation of `IHostedService` that passes the list of tenants to the scheduler repository class and makes the repository create and start them in its `StartAsync()` method. Also remember to shutdown all schedulers in the `StopAsync()` method.
5. Add the hosted service in your `Startup.cs`

This does it for the one-schedulder-per-tenant part. 
Additionally, because we are using Entity Framework database contexts in our Quartz jobs, we needed to supply each job with a context that has the correct connections string associated with it. To do this we created our own `JobFactory` just as @lahma described in his answer.

I hope this helps anyone. 
Thank you for your time.

**Reply by lahma** [reply]
Thank you for sharing your solution and I'm glad that it worked out at the end!
